mmedit/models/backbones/sr_backbones/cain_net.py

```diff
@@ -0,0 +1,386 @@
+
+    Returns:
+        padding_input (Function): Padding function.
+        padding_output (Function): Depadding function.
+    """
+
+    h, w = x.shape[-2:]
+    padding_width, padding_height = 0, 0
+    if w != ((w >> padding) << padding):
+        padding_width = (((w >> padding) + 1) << padding) - w
+    if h != ((h >> padding) << padding):
+        padding_height = (((h >> padding) + 1) << padding) - h
+    left, right = padding_width // 2, padding_width - padding_width // 2
+    up, down = padding_height // 2, padding_height - padding_height // 2
+    # print(up, down, left, right)
+    if down >= h or right >= w:
+        function = nn.ReplicationPad2d
+    else:
+        function = nn.ReflectionPad2d
+    padding_input = function(padding=[left, right, up, down])
+    padding_output = function(padding=[0 - left, 0 - right, 0 - up, 0 - down])
+    return padding_input, padding_output
+
+
+class ConvNorm(nn.Module):
+    """ReflectionPad2d, Conv2d (and Norm).
+
+    Args:
+        in_channels (int): Channel number of input features.
+        out_channels (int): Channel number of input features.
+        kernel_size (int): Kernel size of convolution layer.
+        stride (int): Stride of size of convolution layer. Default: 1.
+        norm (None | function): Norm layer. If None, no norm layer.
+            Default: None.
+    """
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 kernel_size,
+                 stride=1,
+                 norm=None):
+        super().__init__()
+
+        reflection_padding = kernel_size // 2
+        self.reflection_pad = nn.ReflectionPad2d(reflection_padding)
+        self.conv = nn.Conv2d(
+            in_channels,
+            out_channels,
+            stride=stride,
+            kernel_size=kernel_size,
+            bias=True)
+
+        self.norm = norm
+        if norm == 'IN':
+            self.norm = nn.InstanceNorm2d(
+                out_channels, track_running_stats=True)
+        elif norm == 'BN':
+            self.norm = nn.BatchNorm2d(out_channels)
+
+    def forward(self, x):
+        """Forward function for ConvNorm.
+
+        Args:
+            x (Tensor): Input tensor with shape (n, c, h, w).
+
+        Returns:
+            Tensor: Output tensor with shape (n, c, h, w).
+        """
+
+        out = self.reflection_pad(x)
+        out = self.conv(out)
+        if self.norm:
+            out = self.norm(out)
+        return out
+
+
+class CALayer(nn.Module):
+    """Channel Attention Layer.
+
+    Args:
+        mid_channels (int): Channel number of the intermediate features.
+        reduction (int): Channel reduction of CA. Default: 16.
+    """
+
+    def __init__(self, mid_channels, reduction=16):
+        super().__init__()
+
+        # global average pooling: feature --> point
+        self.avg_pool = nn.AdaptiveAvgPool2d(1)
+        # feature channel downscale and upscale --> channel weight
```

Comment:
Not so sure what this means (I understand the operation below, but the comment needs to be modified)